atomix/cluster/src/test/java/io/atomix/raft/RaftRule.java

```diff
@@ -0,0 +1,521 @@
+    final RaftStorage.Builder defaults =
+        RaftStorage.builder()
+            .withStorageLevel(StorageLevel.DISK)
+            .withDirectory(new File(directory.toFile(), memberId.toString()))
+            .withMaxEntriesPerSegment(10)
+            .withMaxSegmentSize(1024 * 10)
+            .withNamespace(RaftNamespaces.RAFT_STORAGE);
+    return configurator.apply(defaults).build();
+  }
+
+  private Optional<RaftServer> getLeader() {
+    return servers.stream().filter(s -> s.getRole() == Role.LEADER).findFirst();
+  }
+
+  private Optional<RaftServer> getFollower() {
+    return servers.stream().filter(s -> s.getRole() == Role.FOLLOWER).findFirst();
+  }
+
+  public void appendEntries(final int count) {
+    final var leader = getLeader().orElseThrow();
+
+    for (int i = 0; i < count; i++) {
+      appendEntryAsync(leader, 1024);
+    }
+  }
+
+  public long appendEntry() throws Exception {
+    final var leader = getLeader().orElseThrow();
+
+    return appendEntry(leader, 1024);
+  }
+
+  private long appendEntry(final RaftServer leader, final int entrySize) throws Exception {
+    final var raftRole = leader.getContext().getRaftRole();
+    if (raftRole instanceof LeaderRole) {
+      final var testAppendListener = appendEntry(entrySize, (LeaderRole) raftRole);
+      return testAppendListener.awaitCommit();
+    }
+    throw new IllegalArgumentException(
+        "Expected to append entry on leader, "
+            + leader.getContext().getName()
+            + " was not the leader!");
+  }
+
+  private void appendEntryAsync(final RaftServer leader, final int entrySize) {
+    final var raftRole = leader.getContext().getRaftRole();
+
+    if (raftRole instanceof LeaderRole) {
+      appendEntry(entrySize, (LeaderRole) raftRole);
+      return;
+    }
+
+    throw new IllegalArgumentException(
+        "Expected to append entry on leader, "
+            + leader.getContext().getName()
+            + " was not the leader!");
+  }
+
+  private TestAppendListener appendEntry(final int entrySize, final LeaderRole leaderRole) {
+    final var appendListener = new TestAppendListener();
+    position += 1;
+    leaderRole.appendEntry(
+        position,
+        position + 10,
+        ByteBuffer.wrap(RandomStringUtils.random(entrySize).getBytes()),
+        appendListener);
+    position += 10;
+    return appendListener;
+  }
+
+  public void awaitAppendEntries(final int i) throws Exception {
+    // this call is async
+    appendEntries(i - 1);
+
+    // this awaits the last append
+    appendEntry();
+  }
+
+  @Override
+  public String toString() {
+    return "RaftRule with " + nodeCount + " nodes.";
+  }
+
+  private static final class CommitAwaiter {
+
+    private final long awaitedIndex;
+    private final CountDownLatch latch = new CountDownLatch(1);
+
+    public CommitAwaiter(final long index) {
+      this.awaitedIndex = index;
+    }
+
+    public boolean reachedCommit(final long index) {
```

Comment:
rename to `reachedCommit(final long currentIndex)` The way it is now it is confusing